mage_ai/data_cleaner/column_type_detector.py

```diff
@@ -54,143 +56,117 @@ def get_mismatched_row_count(series, column_type):
+                mdtype = DATETIME
+            else:
+                correct_emails = clean_series.str.match(REGEX_EMAIL).sum()
+                correct_phone_nums = clean_series.str.match(REGEX_PHONE_NUMBER).sum()
+                correct_zip_codes = clean_series.str.match(REGEX_ZIP_CODE).sum()
+                if correct_emails/length >= 0.90:
+                    mdtype = EMAIL
+                elif correct_phone_nums/length >= 0.90:
+                    mdtype = PHONE_NUMBER
+                elif correct_zip_codes/length >= 0.90:
+                    mdtype = ZIP_CODE
+                elif series_unique == 2:
+                    mdtype = TRUE_OR_FALSE
                 else:
-                    category_feature_names.append(col_name)
-
-    low_cardinality_category_feature_names = \
-        [col for col in category_feature_names if df[col].nunique() <= kwargs.get(
-            'category_cardinality_threshold',
-            255,
-        )]
-    high_cardinality_category_feature_names = \
-        [col for col in category_feature_names if col not in low_cardinality_category_feature_names]
-
-    column_types = {}
-    array_types_mapping = {
-        CATEGORY: low_cardinality_category_feature_names,
-        CATEGORY_HIGH_CARDINALITY: high_cardinality_category_feature_names,
-        DATETIME: datetime_feature_names,
-        EMAIL: email_features,
-        NUMBER: integer_feature_names,
-        NUMBER_WITH_DECIMALS: float_feature_names,
-        PHONE_NUMBER: phone_number_feature_names,
-        TEXT: text_feature_names,
-        TRUE_OR_FALSE: binary_feature_names,
-        ZIP_CODE: zip_code_feature_names,
-    }
-    for col_type, arr in array_types_mapping.items():
-        for col in arr:
-            column_types[col] = col_type
+                    if type(exact_dtype) is list:
+                        mdtype = TEXT
+                    elif clean_series_unique / length >= 0.8:
+                        mdtype = TEXT
+                    else:
+                        word_count = clean_series.map(lambda x: len(str(x).split(' '))).max()
+                        if word_count > MAXIMUM_WORD_LENGTH_FOR_CATEGORY_FEATURES:
+                            mdtype = TEXT
+                        else:
+                            if (clean_series_unique <= kwargs.get(
+                                'category_cardinality_threshold', 255)):
+                                mdtype =  CATEGORY
+                            else:
+                                mdtype = CATEGORY_HIGH_CARDINALITY
+    return mdtype
 
-    return column_types
+def infer_column_types(df, **kwargs):
+    columns = [df[col] for col in df.columns]
+    column_names = list(df.columns)
+    dtypes = list(df.dtypes)
+    kwarg_list = [kwargs] * len(columns)
+    ctypes = {}
+    num_entries = len(df)
+    if num_entries > MULTITHREAD_MAX_NUM_ENTRIES:
+        with ThreadPoolExecutor(max_workers=MAX_WORKERS) as pool:
```

Comment:
Let's add the multi thread method to https://github.com/mage-ai/mage-ai/blob/master/mage_ai/data_cleaner/shared/multi.py and use here

Reply:
Added